posix/consolefs.c

```diff
@@ -0,0 +1,295 @@
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        if (retval == -1)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    /* Allocate and initialize a new file structure. */
+    {
+        file_t* new_file;
+
+        if (!(new_file = oe_calloc(1, sizeof(file_t))))
+            OE_RAISE_ERRNO(OE_ENOMEM);
+
+        *new_file = *file;
+        new_file->host_fd = retval;
+        *new_file_out = (oe_device_t*)new_file;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static int _consolefs_ioctl(
+    oe_device_t* file,
+    unsigned long request,
+    uint64_t arg)
+{
+    int ret = -1;
+
+    OE_UNUSED(file);
+    OE_UNUSED(request);
+    OE_UNUSED(arg);
+    OE_RAISE_ERRNO(OE_ENOTSUP);
+
+done:
+    return ret;
+}
+
+static int _consolefs_fcntl(oe_device_t* file, int cmd, uint64_t arg)
+{
+    int ret = -1;
+
+    OE_UNUSED(file);
+    OE_UNUSED(cmd);
+    OE_UNUSED(arg);
+    OE_RAISE_ERRNO(OE_ENOTSUP);
+
+done:
+    return ret;
+}
+
+static ssize_t _consolefs_read(oe_device_t* file_, void* buf, size_t count)
+{
+    ssize_t ret = -1;
+    file_t* file = _cast_file(file_);
+
+    oe_errno = 0;
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_read_ocall(&ret, file->host_fd, buf, count) != OE_OK)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+done:
+    return ret;
+}
+
+static ssize_t _consolefs_write(
+    oe_device_t* file_,
+    const void* buf,
+    size_t count)
+{
+    ssize_t ret = -1;
+    file_t* file = _cast_file(file_);
+
+    oe_errno = 0;
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_write_ocall(&ret, file->host_fd, buf, count) != OE_OK)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+done:
+    return ret;
+}
+
+static oe_host_fd_t _consolefs_gethostfd(oe_device_t* file_)
+{
+    oe_host_fd_t ret = -1;
+    file_t* file = _cast_file(file_);
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    ret = file->host_fd;
+
+done:
+    return ret;
+}
+
+static oe_off_t _consolefs_lseek(
+    oe_device_t* file_,
+    oe_off_t offset,
+    int whence)
+{
+    oe_off_t ret = -1;
+    file_t* file = _cast_file(file_);
+
+    oe_errno = 0;
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_lseek_ocall(&ret, file->host_fd, offset, whence) != OE_OK)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+done:
+    return ret;
+}
+
+static int _consolefs_close(oe_device_t* file_)
+{
+    int ret = -1;
+    file_t* file = _cast_file(file_);
+    oe_errno = 0;
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Ask the host to perform this operation. */
+    {
+        if (oe_posix_close_ocall(&ret, file->host_fd) != OE_OK)
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        if (ret == -1)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    switch (file->host_fd)
+    {
+        case OE_STDIN_FILENO:
+        case OE_STDOUT_FILENO:
+        case OE_STDERR_FILENO:
+        {
+            /* Do not free these statically initialized structures. */
+            file->host_fd = -1;
```

Comment:
This doesn't seem like a definitive check for whether `file` is statically initialized. Is this an issue?

```
dup2(OE_STDIN_FILENO, tmp);
close(OE_STDIN_FILENO); // Static version is invalidated.
dup2(tmp, OE_STDIN_FILENO);
close(tmp); // first dupe is freed
close(OE_STDIN_FILENO); // second dupe is invalidated but not freed?
```

Reply:
Okay, I reworked consolefs and added dup2() tests (including the above). The standard devices are now heap allocated like everything else.